usr/share/rear/lib/config-functions.sh

```diff
@@ -16,15 +16,60 @@ function SetOSVendorAndVersion () {
     # The test must match OS_VENDOR=generic or OS_VERSION=none in default.conf:
     if test "$OS_VENDOR" = generic -o "$OS_VERSION" = none ; then
 
-        # Recent Linux distro's with systemd has the /etc/os-release file
+        # Recent Linux distros with systemd has the /etc/os-release file
         # Try to find all the required information from that file
         if [[ -f /etc/os-release ]] ; then
-            grep -q -i 'fedora' /etc/os-release && OS_VENDOR=Fedora
-            grep -q -i -E '(centos|redhat|scientific|oracle)' /etc/os-release && OS_VENDOR=RedHatEnterpriseServer
-            grep -q -i 'suse' /etc/os-release && OS_VENDOR=SUSE_LINUX
-            grep -q -i 'debian' /etc/os-release && OS_VENDOR=Debian
-            grep -q -i -E '(ubuntu|linuxmint)' /etc/os-release && OS_VENDOR=Ubuntu
-            grep -q -i 'arch' /etc/os-release && OS_VENDOR=Arch
+            local ID ID_LIKE
+            eval "$(grep '^ID=' /etc/os-release)"
```

Comment:
Sorry for being late to the party, I know that the same issue was discussed already in PR #3165 ( https://github.com/rear/rear/pull/3165#discussion_r1503932927 ) and I realize that a simple method of obtaining a variable assignment from a file with shell syntax (using `source`) has a problem when the file contains some variables that are `readonly`. Why not use another shell call instead of a `(  )` subshell though? Readonly variables will not propagate through the execution:
```
$ readonly VERSION_ID
$ bash -c 'source /etc/os-release; echo "$VERSION_ID"'
38
$ function get_var_from_file() { bash -c 'source "$0"; echo "${!1}"' "$1" "$2"; }
$ get_var_from_file /etc/os-release VERSION_ID
38
```
(the `get_var_from_file` function is meant to be generic and safe against whitespace and special characters in the file name).

I think this is simpler and more robust than using `eval` and `grep` (think of possible multi-line strings, for example).

Reply:
@pcahyna 
welcome late to the party!

A perfect example of Linus's law
```
given enough eyeballs, all issues are shallow
```
(I replaced here "bugs" with the more generic "issues").

Reply:
@lzaoral 
when you implement the get_var_from_file() function
please make a verbose comment that explains WHY
it is implemented as it is to avoid that
at any time later others may think the
explicit bash in between is superfluous
and it could be "just simplified" ("KISSed")
to something insufficient or unreliable
as what we had tried before here.

That explicit bash in between is a nice example
that what RFC 1925 item 6a means is not always right
so an additional RFC 1925 item 6b could be needed like
```
(6b) (contrary).
Sometimes another level of indirection is mandatory.
```
which is basically what I mean with KSIS
"Keep Separated Issues Separated"
that is somewhat related to RFC 1925 item 5
so in the end RFC 1925 is already complete
when understood "cum grano salis" ;-)

Reply:
Thank you both!  Implemented in a0e53eddd6f5d621cde93f49efebd7f127339344.

Reply:
@lzaoral 
because "the get_var_from_file function is meant to be generic"
cf. https://github.com/rear/rear/pull/3171#discussion_r1521750947
I think it fits perhaps better in lib/global-functions.sh
therein e.g. next to the function read_and_strip_file()

Reply:
Just a note: this method will not be able to read Bash arrays. In this case it is fine, because os-release it meant to be readable by all Bourne-compatible shells, not just Bash, and does not contain arrays, but should be kept in mind. In particular one should not use this method to read array variables from ReaR's configuration files.

Reply:
Proposal for a (hopefully sufficiently) generic function
get_assignment_from_file()
that also works (at least for my current tests)
to get arrays from a file
(but I needed that problematic 'eval'):

```
# cat my-shell-assignments
VAR="value"
STRING="words \
in \
a \
string"
ARR=( 'first element'
      'second element'
      'last element' )

# function get_assignment_from_file() { bash -c 'source "$0" ; declare -p "$1" | cut -d " " -f3- | sed -e "s/^$1/$2/"' "$1" "$2" "$3" ; }

# readonly VAR STRING ARR

# eval "$( get_assignment_from_file my-shell-assignments VAR myvar )"

# eval "$( get_assignment_from_file my-shell-assignments STRING mystring )"

# eval "$( get_assignment_from_file my-shell-assignments ARR myarr )"

# declare -p myvar mystring myarr
declare -- myvar="value"
declare -- mystring="words in a string"
declare -a myarr=([0]="first element" [1]="second element" [2]="last element")
```

Reply:
Simpler version of a (hopefully sufficiently) generic function
`get_rvalue_from_file()`
to get rvalues (right hand side of an assignment)
that also works (at least for my current tests)
to get the rvalue for arrays from a file.
It still needs the problematic 'eval'
plus a simpler 'cut' but no longer 'sed'.
```
# cat my-shell-assignments
VAR="value"
STRING="lvalue \
= \
rvalue"
ARR=( 'first element = 1'
      'second element = 2'
      'last element = 3' )

# function get_rvalue_from_file() { bash -c 'source "$0" ; declare -p "$1" | cut -s -d "=" -f2-' "$1" "$2" ; }

# readonly VAR STRING ARR

# eval myvar=$( get_rvalue_from_file my-shell-assignments VAR )

# eval mystring=$( get_rvalue_from_file my-shell-assignments STRING )
# eval myarr=$( get_rvalue_from_file my-shell-assignments ARR )

# declare -p myvar mystring myarr
declare -- myvar="value"
declare -- mystring="lvalue = rvalue"
declare -a myarr=([0]="first element = 1" [1]="second element = 2" [2]="last element = 3")
```

Reply:
@jsmeix this is interesting, but I find it more fragile than my version, so I would not use it until we actually need to read files with arrays (and even then we should have both versions, one for arrays, one without).
Regarding `eval`: that's how the output of `declare -p` is meant to be used, I don't think using it in another way will be correct.

Reply:
Yes, it is more complicated and therefore more fragile and
it is more complicated to be used (also needs 'eval').
I did it because I was curious how it might be implemented.
But as long as only a simple variable is needed
the simpler get_var_from_file() should be preferred.

Reply:
I like to leve this comment open
so that at any time later one (at least I)
can see directly what I tried out here with
a function like get_rvalue_from_file()
to get rvalues (right hand side of an assignment)
that also works (at least for my current tests)
to get the rvalue for arrays from a "sourceable" file.